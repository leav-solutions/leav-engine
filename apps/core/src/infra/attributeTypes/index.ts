// Copyright LEAV Solutions 2017 until 2023/11/05, Copyright Aristid from 2023/11/06
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export {default as attributeAdvancedLink} from './attributeAdvancedLinkRepo';
export {default as attributeAdvanced} from './attributeAdvancedRepo';
export {default as attributeSimpleLink} from './attributeSimpleLinkRepo';
export {default as attributeSimple} from './attributeSimpleRepo';
export {default as attributeTree} from './attributeTreeRepo';
export {default} from './attributeTypesRepo';
