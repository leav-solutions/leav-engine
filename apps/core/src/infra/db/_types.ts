// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
// Copyright LEAV Solutions 2017
// This file is released under LGPL V3

import {AqlQuery} from 'arangojs/lib/cjs/aql-query';
import {IQueryInfos} from '_types/queryInfos';

// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export interface IDbDocument {
    _id: string;
    _key: string;
    _rev: string;
    [key: string]: any;
}

export interface IDbEdge extends IDbDocument {
    _from: string;
    _to: string;
}

export interface IExecute {
    query: string | AqlQuery;
    ctx: IQueryInfos;
    withTotalCount?: boolean;
    attempts?: number;
}

export interface IExecuteWithCount<T = IDbDocument> {
    totalCount: number;
    results: T[];
}

export const isExecuteWithCount = (res: IExecuteWithCount<unknown> | unknown[]): res is IExecuteWithCount<unknown> => {
    return typeof (res as IExecuteWithCount).results !== 'undefined';
};
