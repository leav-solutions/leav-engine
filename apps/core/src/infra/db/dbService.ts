// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {getCallStack} from '@leav/utils';
import {Database} from 'arangojs';
import {isAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {createHash} from 'crypto';
import {IUtils} from 'utils/utils';
import {IConfig} from '_types/config';
import {IDbProfiler} from '_types/dbProfiler';
import {IDbDocument, IExecute, IExecuteWithCount} from './_types';

const MAX_ATTEMPTS = 10;

export interface IDbService {
    db?: Database;

    /**
     * Execute an AQL query
     * If withTotalCount is set to true, return an object with totalCount and results. Otherwise, just return
     * results straight from DB.
     *
     * @param query
     * @param withTotalCount
     * @param attempts Used when we have to retry a query after a write-write conflict
     * @throws If query fails or we still have a conflict after all attempts
     */
    execute?<T extends IExecuteWithCount<unknown> | unknown[] = IDbDocument[]>(params: IExecute): Promise<T>;

    /**
     * Create a new collection in database
     *
     * @param name Collection name
     * @param type Document or edge collection?
     * @throws If collection already exists
     */
    createCollection?(name: string, type?: collectionTypes): Promise<void> | null;

    /**
     * Delete a collection in database
     *
     * @param name Collection name
     * @param type Document or edge collection?
     * @throws If collection already exists
     */
    dropCollection?(name: string, type?: collectionTypes): Promise<void> | null;

    /**
     * Check if collection already exists
     *
     * @param name
     */
    collectionExists?(name: string): Promise<boolean>;
}

export enum collectionTypes {
    DOCUMENT = 'document',
    EDGE = 'edge'
}

interface IDeps {
    'core.infra.db'?: Database;
    'core.utils'?: IUtils;
    config?: IConfig;
}

export default function ({
    'core.infra.db': db = null,
    'core.utils': utils = null,
    config = null
}: IDeps = {}): IDbService {
    const collectionExists = async function (name: string): Promise<boolean> {
        const collections = await db.listCollections();

        return collections.reduce((exists, c) => exists || c.name === name, false);
    };

    const _sleep = (ms: number): Promise<void> => {
        return new Promise(resolve => setTimeout(resolve, ms));
    };

    return {
        db,
        async execute<T extends IExecuteWithCount<any> | any[] = any[]>({
            query,
            ctx,
            withTotalCount = false,
            attempts = 0
        }: IExecute): Promise<T> {
            try {
                if (config.dbProfiler.enable) {
                    const dbProfiler: IDbProfiler = ctx.dbProfiler ?? {
                        totalCount: 0,
                        uniqueQueriesCount: 0,
                        queries: {}
                    };

                    dbProfiler.totalCount = (dbProfiler.totalCount ?? 0) + 1;

                    // Generate a hash from the query to be able
                    // to group identical queries (exact same query with exact same params)
                    const queryKey = createHash('md5').update(JSON.stringify(query)).digest('base64');

                    if (!dbProfiler.queries) {
                        dbProfiler.queries = {};
                    }

                    const callStack = JSON.stringify(getCallStack(10));

                    dbProfiler.queries[queryKey] = {
                        count: (dbProfiler.queries?.[queryKey]?.count ?? 0) + 1,
                        callers: (dbProfiler.queries?.[queryKey]?.callers ?? new Set()).add(callStack),
                        query
                    };

                    dbProfiler.uniqueQueriesCount = Object.keys(dbProfiler.queries).length;

                    ctx.dbProfiler = dbProfiler;
                }

                // Convert query to AqlQuery if we have a simple query to match query() types
                const queryToRun = isAqlQuery(query)
                    ? {...query}
                    : {
                          query,
                          bindVars: {}
                      };

                const queryOptions = withTotalCount ? {count: true, options: {fullCount: true}} : {};

                const cursor = await db.query(queryToRun, queryOptions);

                const results = await cursor.all();
                return withTotalCount ? {totalCount: cursor.extra.stats.fullCount, results} : results;
            } catch (e) {
                // Handle write-write conflicts: we try the query again with a growing delay between trials.
                // If we reach maximum attempts and still no success, stop it and throw the exception
                // error 1200 === conflict
                if (e.isArangoError && e.errorNum === 1200 && attempts < MAX_ATTEMPTS) {
                    const timeToWait = 2 ** attempts;
                    await _sleep(timeToWait);
                    return this.execute({query, ctx, withTotalCount, attempts: attempts + 1});
                }

                e.message += `\nQuery was: ${JSON.stringify(query).replace(/\\n/g, ' ')}`;
                e.query = query;

                // Response contains circular references which can cause an error when converted to JSON
                // later on. It doesn't contains useful information anyway, so throw it away.
                delete e.response;

                utils.rethrow(e);
            }
        },
        async createCollection(name: string, type = collectionTypes.DOCUMENT): Promise<void> {
            if (await collectionExists(name)) {
                throw new Error(`Collection ${name} already exists`);
            }

            if (type === collectionTypes.EDGE) {
                const collection = db.edgeCollection(name);
                await collection.create();
            } else {
                const collection = db.collection(name);
                await collection.create();
            }
        },
        async dropCollection(name: string, type = collectionTypes.DOCUMENT): Promise<void> {
            if (!(await collectionExists(name))) {
                throw new Error(`Collection ${name} does not exist`);
            }

            const collection = type === collectionTypes.EDGE ? db.edgeCollection(name) : db.collection(name);
            await collection.drop();
        },
        collectionExists
    };
}
