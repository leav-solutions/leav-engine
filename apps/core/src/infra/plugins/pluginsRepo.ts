// Copyright LEAV Solutions 2017 until 2023/11/05, Copyright Aristid from 2023/11/06
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IPluginInfos, IRegisteredPlugin} from '_types/plugin';

export interface IPluginsRepo {
    registerPlugin(path: string, plugin: IPluginInfos): IRegisteredPlugin;
    getRegisteredPlugins(): IRegisteredPlugin[];
}

export default function (): IPluginsRepo {
    const _registeredPlugins: IRegisteredPlugin[] = [];

    return {
        registerPlugin(path: string, plugin: IPluginInfos): IRegisteredPlugin {
            const pluginToRegister = {path, infos: {...plugin}};
            _registeredPlugins.push(pluginToRegister);

            return pluginToRegister;
        },
        getRegisteredPlugins(): IRegisteredPlugin[] {
            return _registeredPlugins.sort((a, b) => a.infos.name.localeCompare(b.infos.name));
        }
    };
}
