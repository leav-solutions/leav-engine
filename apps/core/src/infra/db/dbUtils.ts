// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql} from 'arangojs';
import {GeneratedAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {CollectionType} from 'arangojs/lib/cjs/collection';
import {AwilixContainer} from 'awilix';
import {accessSync, constants, readdirSync} from 'fs';
import {IPluginsRepo} from 'infra/plugins/pluginsRepo';
import * as path from 'path';
import * as winston from 'winston';
import {IAttribute} from '_types/attribute';
import {IConfig} from '_types/config';
import {ILibrary} from '_types/library';
import {IList, IPaginationParams, ISortParams} from '_types/list';
import {IQueryInfos} from '_types/queryInfos';
import {IKeyValue} from '_types/shared';
import {ITree} from '_types/tree';
import {IDbValueVersion, IValueVersion} from '_types/value';
import {ECacheType, ICachesService} from '../../infra/cache/cacheService';
import {collectionTypes, IDbService} from './dbService';
import runMigrationFiles from './helpers/runMigrationFiles';
import {IExecuteWithCount} from './_types';

export const MIGRATIONS_COLLECTION_NAME = 'core_db_migrations';

export type CustomFilterConditionsFunc = (
    filterKey: string,
    filterVal: string | boolean | string[],
    strictFilters: boolean
) => GeneratedAqlQuery;

export interface IFindCoreEntityParams {
    collectionName: string;
    filters?: ICoreEntityFilterOptions;
    strictFilters?: boolean;
    withCount?: boolean;
    pagination?: IPaginationParams;
    sort?: ISortParams;
    customFilterConditions?: IKeyValue<CustomFilterConditionsFunc>;
    nonStrictFields?: string[];
    ctx: IQueryInfos;
}

export interface IDbUtils {
    migrate?(depsManager: AwilixContainer): Promise<void>;
    cleanup?<T extends {}>(record: {}): T;
    convertToDoc?(obj: {}): any;
    isCollectionExists?(name: string): Promise<boolean>;
    findCoreEntity?<T extends ICoreEntity>(params: IFindCoreEntityParams): Promise<IList<T>>;
    clearDatabase?(): Promise<void>;
}

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.infra.cache.cacheService'?: ICachesService;
    'core.utils.logger'?: winston.Winston;
    'core.infra.plugins'?: IPluginsRepo;
    config?: IConfig;
}

export default function ({
    'core.infra.db.dbService': dbService = null,
    'core.infra.cache.cacheService': cacheService = null,
    'core.utils.logger': logger = null,
    'core.infra.plugins': pluginsRepo = null,
    config = null
}: IDeps = {}): IDbUtils {
    /**
     * Create the collections used to managed db migrations
     * This can't be in a migration file because we do have to initialize it somewhere
     *
     */
    async function _initMigrationsCollection(): Promise<void> {
        const collections = await dbService.db.listCollections();

        const colExists = collections.reduce((exists, c) => exists || c.name === MIGRATIONS_COLLECTION_NAME, false);
        if (!colExists) {
            const collection = dbService.db.collection(MIGRATIONS_COLLECTION_NAME);
            await collection.create();
        }
    }

    /**
     * Return the filter's conditions based on key and val supplied.
     *
     * @param filterKey
     * @param filterVal
     * @param bindVars
     * @param index
     * @param strictFilters
     */
    function _getFilterCondition(
        filterKey: string,
        filterVal: string | boolean | string[],
        strictFilters: boolean,
        nonStrictFields?: string[]
    ): GeneratedAqlQuery {
        const queryParts = [];

        // If value is an array (types or formats for example),
        // we call this function recursively on array and join filters with an OR
        if (Array.isArray(filterVal)) {
            if (filterVal.length) {
                const valParts = filterVal.map(val => _getFilterCondition(filterKey, val, strictFilters));
                queryParts.push(aql.join(valParts, ' OR '));
            }
        } else {
            if (filterKey === 'label') {
                // Search for label in any language
                const valParts = config.lang.available.map(l => aql`LIKE(el.label.${l}, ${filterVal}, true)`);
                queryParts.push(aql.join(valParts, ' OR '));
            } else {
                // Filter with a "like" on ID or exact value in other fields
                queryParts.push(
                    (nonStrictFields ?? []).includes(filterKey) && !strictFilters
                        ? aql`LIKE(el.${filterKey}, ${filterVal}, true)`
                        : aql`el.${filterKey} == ${filterVal}`
                );
            }
        }

        return aql.join(queryParts);
    }

    const ret = {
        /**
         * Run database migrations.
         * It takes all files present in migrations folder and run it if it's never been executed before
         *
         * @param depsManager
         */
        async migrate(depsManager: AwilixContainer): Promise<void> {
            await _initMigrationsCollection();
            const ctx: IQueryInfos = {
                userId: config.defaultUserId,
                queryId: 'run-migrations'
            };
            // Load already ran migrations
            const executedMigrations = await dbService.execute<string[]>({
                query: `
                    FOR m IN core_db_migrations
                    RETURN m.file
                `,
                ctx
            });

            const _runMigrationFiles = (files, folder, prefix = null) =>
                runMigrationFiles({
                    files,
                    executedMigrations,
                    migrationsDir: folder,
                    prefix,
                    deps: {depsManager, dbService, logger},
                    ctx
                });

            /*** Core migrations ***/
            // Load migrations files
            const migrationsDir = path.resolve(__dirname, 'migrations');
            const migrationFiles = readdirSync(migrationsDir).filter(file => file.indexOf('.map') === -1);

            await _runMigrationFiles(migrationFiles, migrationsDir);

            /*** Plugins migrations ***/
            const plugins = pluginsRepo.getRegisteredPlugins();
            for (const plugin of plugins) {
                const pluginMigrationFolder = path.resolve(plugin.path + '/infra/db/migrations');

                try {
                    accessSync(pluginMigrationFolder, constants.R_OK);
                } catch (e) {
                    continue;
                }

                const pluginMigrationFiles = readdirSync(pluginMigrationFolder).filter(
                    file => file.indexOf('.map') === -1
                );

                await _runMigrationFiles(pluginMigrationFiles, pluginMigrationFolder, plugin.infos.name);
            }

            /** Clear cache */
            for (const cacheType of Object.values(ECacheType)) {
                cacheService.getCache(cacheType).deleteAll();
            }
        },

        /**
         * Cleanup every system keys from an object coming from database.
         * _key is kept under 'id'
         *
         * @param obj
         * @return any   Cleaned up object
         */
        cleanup(obj: any): any {
            if (obj === null || typeof obj === 'undefined') {
                return null;
            }

            return Object.keys(obj).reduce((newObj: any, key) => {
                if (key === '_key') {
                    newObj.id = obj[key];
                } else if (key[0] !== '_') {
                    newObj[key] = obj[key];
                }

                return newObj;
            }, {});
        },

        /**
         * Convert an object to an object looking like a DB document
         * id is replaced by _key
         *
         * @param obj
         * @return any   DB document compatible object
         */
        convertToDoc(obj: {}): any {
            const newObj: any = {...obj};

            if (typeof newObj.id !== 'undefined') {
                newObj._key = newObj.id;
            }
            delete newObj.id;

            return newObj;
        },
        /**
         * Search core entities (libraries, attributes, trees)
         *
         * @param collectionName
         * @param filters
         * @param strictFilters
         */
        async findCoreEntity<T extends ITree | ILibrary | IAttribute>(
            params: IFindCoreEntityParams
        ): Promise<IList<T>> {
            const {
                collectionName = null,
                filters = null,
                strictFilters = false,
                withCount = false,
                pagination = null,
                sort = null,
                customFilterConditions = {},
                nonStrictFields = ['label', '_key'],
                ctx = {}
            } = params;

            const collec = dbService.db.collection(collectionName);
            const queryParts = [aql`FOR el IN ${collec}`];

            if (filters !== null) {
                const dbFilters = ret.convertToDoc(filters);
                const filtersKeys = Object.keys(dbFilters);

                for (const filterKey of filtersKeys) {
                    const filterVal = dbFilters[filterKey];

                    // Caller can define some custom functions to generate filter condition (like looking on edges to
                    // filter on libraries linked to an attribute). So, if a custom function is define for this filter,
                    // we use it, otherwise we use the standard filters
                    const filterCondsFunc =
                        typeof customFilterConditions[filterKey] !== 'undefined'
                            ? customFilterConditions[filterKey]
                            : _getFilterCondition;
                    const conds = filterCondsFunc(filterKey, filterVal, strictFilters, nonStrictFields);

                    if (conds?.query) {
                        queryParts.push(aql`FILTER`, conds);
                    }
                }
            }

            if (!!sort) {
                const field = sort.field === 'id' ? '_key' : sort.field;
                queryParts.push(aql`SORT el.${field} ${sort.order}`);
            }

            if (!!pagination) {
                queryParts.push(aql`LIMIT ${pagination.offset || 0}, ${pagination.limit}`);
            }

            queryParts.push(aql`RETURN el`);

            const query = aql.join(queryParts);
            const res = await dbService.execute<IExecuteWithCount | any[]>({query, withTotalCount: withCount, ctx});

            const results = !Array.isArray(res) ? res.results : res;

            return {
                totalCount: withCount ? (res as IExecuteWithCount).totalCount : null,
                list: results.map(ret.cleanup)
            };
        },
        convertValueVersionToDb(version: IValueVersion): IDbValueVersion {
            return Object.keys(version).reduce((allVers, treeName) => {
                const id = version[treeName];

                allVers[treeName] = id;

                return allVers;
            }, {});
        },
        async clearDatabase(): Promise<void> {
            // Drop all collections
            const cols = await dbService.db.listCollections();
            for (const col of cols) {
                const colType =
                    col.type === CollectionType.DOCUMENT_COLLECTION ? collectionTypes.DOCUMENT : collectionTypes.EDGE;
                await dbService.dropCollection(col.name, colType);
            }
        }
    };

    return ret;
}
