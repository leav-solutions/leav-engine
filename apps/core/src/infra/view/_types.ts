// Copyright LEAV Solutions 2017 until 2023/11/05, Copyright Aristid from 2023/11/06
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IDbService} from 'infra/db/dbService';
import {IDbUtils} from 'infra/db/dbUtils';
import {IList} from '_types/list';
import {IGetCoreEntitiesParams} from '_types/shared';
import {IQueryInfos} from '../../_types/queryInfos';

export const VIEWS_COLLECTION_NAME = 'core_views';

export interface IViewRepo {
    createView(view: IView, ctx: IQueryInfos): Promise<IView>;
    updateView(view: IView, ctx: IQueryInfos): Promise<IView>;
    getViews(params: IGetCoreEntitiesParams, ctx: IQueryInfos): Promise<IList<IView>>;
    deleteView(viewId: string, ctx: IQueryInfos): Promise<IView>;
}

export interface IViewRepoDeps {
    'core.infra.db.dbService': IDbService;
    'core.infra.db.dbUtils': IDbUtils;
}
