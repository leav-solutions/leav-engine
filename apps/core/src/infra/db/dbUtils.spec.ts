// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql, Database, DocumentCollection} from 'arangojs';
import {asFunction, AwilixContainer} from 'awilix';
import {resolve} from 'dns';
import {readdirSync} from 'fs';
import {IPluginsRepo} from 'infra/plugins/pluginsRepo';
import {Winston} from 'winston';
import {IAttributeFilterOptions} from '_types/attribute';
import {IConfig} from '_types/config';
import {ITree} from '_types/tree';
import {SortOrder} from '../../_types/list';
import {ATTRIB_COLLECTION_NAME} from '../attributeTypes/attributeTypesRepo';
import {TREES_COLLECTION_NAME} from '../tree/treeRepo';
import {IDbService} from './dbService';
import dbUtils, {IDbUtils} from './dbUtils';
import loadMigrationFile from './helpers/loadMigrationFile';

describe('dbUtils', () => {
    const mockConf: Partial<IConfig> = {lang: {available: ['fr', 'en'], default: 'fr'}, defaultUserId: '1'};
    const ctx = {
        userId: '0',
        queryId: '123456'
    };
    afterAll(() => {
        jest.clearAllMocks();
    });

    describe('cleanupSystemKeys', () => {
        test('Should remove all system keys', () => {
            const testDbUtils = dbUtils();

            const testObj = {
                _key: 'testKey',
                _id: 'testId',
                _rev: 'testRev',
                _randomSystemKey: 'test',
                normalKey: 'shouldBeKept'
            };

            const res = testDbUtils.cleanup(testObj);

            expect(res).toMatchObject({id: 'testKey', normalKey: 'shouldBeKept'});
        });

        test('Should return null if param is null', () => {
            const testDbUtils = dbUtils();

            const res = testDbUtils.cleanup(null);

            expect(res).toBeNull();
        });
    });

    describe('convertToDoc', () => {
        test('Should add needed system keys', () => {
            const testDbUtils = dbUtils();

            const testObj = {
                id: 'testId',
                normalKey: 'shouldBeKept'
            };

            const res = testDbUtils.convertToDoc(testObj);

            expect(res).toMatchObject({_key: 'testId', normalKey: 'shouldBeKept'});
        });
    });

    describe('findCoreEntity', () => {
        let mockDbServ;
        let testDbUtils: IDbUtils;
        beforeEach(() => {
            mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([
                    {
                        _key: 'categories',
                        _id: 'core_trees/categories',
                        _rev: '_Wm_Qdtu--_',
                        label: {
                            fr: 'Arbre des catégories'
                        },
                        libraries: ['categories'],
                        system: false
                    }
                ])
            };
            testDbUtils = dbUtils({
                'core.infra.db.dbService': mockDbServ,
                config: mockConf as IConfig
            });
            testDbUtils.cleanup = jest.fn().mockReturnValue({
                id: 'categories',
                system: false,
                label: {
                    fr: 'Arbre des catégories'
                }
            });
            testDbUtils.convertToDoc = jest.fn().mockReturnValue({
                _key: 'categories',
                system: false,
                label: 'Arbre des catégories'
            });
        });

        test('Find core entity without filters', async () => {
            const res = await testDbUtils.findCoreEntity({
                collectionName: TREES_COLLECTION_NAME,
                ctx
            });

            expect(res.list).toHaveLength(1);

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars['@value0']).toBe('core_trees');
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(res.list[0]).toMatchObject({
                id: 'categories',
                system: false,
                label: {
                    fr: 'Arbre des catégories'
                }
            });
        });

        test('Filter with a LIKE on ID', async function () {
            const res = await testDbUtils.findCoreEntity({
                collectionName: TREES_COLLECTION_NAME,
                filters: {id: 'test'},
                ctx
            });

            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/(FILTER LIKE){1}/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();
        });

        test('Should filter label on any language', async function () {
            const res = await testDbUtils.findCoreEntity({
                collectionName: TREES_COLLECTION_NAME,
                filters: {label: 'test'},
                ctx
            });

            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/(LIKE(.*)label\.(.*)OR LIKE(.*)label\.)/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();
        });

        test('Should limit results', async function () {
            const mockDbServLimit = {
                db: new Database(),
                execute: global.__mockPromise([
                    {
                        _key: 'categories',
                        _id: 'core_trees/categories',
                        _rev: '_Wm_Qdtu--_',
                        label: {
                            fr: 'Arbre des catégories'
                        },
                        libraries: ['categories'],
                        system: false
                    }
                ])
            };
            const testDbUtilsLimit = dbUtils({
                'core.infra.db.dbService': mockDbServLimit,
                config: mockConf as IConfig
            });
            const res = await testDbUtilsLimit.findCoreEntity({
                collectionName: TREES_COLLECTION_NAME,
                withCount: true,
                pagination: {limit: 5, offset: 0},
                ctx
            });

            expect(mockDbServLimit.execute.mock.calls[0][0].query.query).toMatch(/LIMIT/);
            expect(mockDbServLimit.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServLimit.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();
        });

        test('Should sort results', async function () {
            const mockDbServLimit = {
                db: new Database(),
                execute: global.__mockPromise([
                    {
                        _key: 'categories',
                        _id: 'core_trees/categories',
                        _rev: '_Wm_Qdtu--_',
                        label: {
                            fr: 'Arbre des catégories'
                        },
                        libraries: ['categories'],
                        system: false
                    }
                ])
            };
            const testDbUtilsLimit = dbUtils({
                'core.infra.db.dbService': mockDbServLimit,
                config: mockConf as IConfig
            });
            const res = await testDbUtilsLimit.findCoreEntity({
                collectionName: TREES_COLLECTION_NAME,
                withCount: true,
                sort: {
                    field: 'system',
                    order: SortOrder.ASC
                },
                ctx
            });

            expect(mockDbServLimit.execute.mock.calls[0][0].query.query).toMatch(/SORT/);
            expect(mockDbServLimit.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServLimit.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();
        });

        test('Should convert ID key when sorting', async function () {
            const mockDbServLimit = {
                db: new Database(),
                execute: global.__mockPromise([
                    {
                        _key: 'categories',
                        _id: 'core_trees/categories',
                        _rev: '_Wm_Qdtu--_',
                        label: {
                            fr: 'Arbre des catégories'
                        },
                        libraries: ['categories'],
                        system: false
                    }
                ])
            };
            const testDbUtilsLimit = dbUtils({
                'core.infra.db.dbService': mockDbServLimit,
                config: mockConf as IConfig
            });
            const res = await testDbUtilsLimit.findCoreEntity({
                collectionName: TREES_COLLECTION_NAME,
                withCount: true,
                sort: {
                    field: 'id',
                    order: SortOrder.ASC
                },
                ctx
            });

            expect(mockDbServLimit.execute.mock.calls[0][0].query.bindVars.value1).toBe('_key');
        });

        test('Should return an empty array if no results', async function () {
            mockDbServ = {db: new Database(), execute: global.__mockPromise([])};
            testDbUtils = dbUtils({
                'core.infra.db.dbService': mockDbServ,
                config: mockConf as IConfig
            });
            testDbUtils.cleanup = jest.fn();
            testDbUtils.convertToDoc = jest.fn();

            mockDbServ = {db: null, execute: global.__mockPromise([])};
            const res = await testDbUtils.findCoreEntity<ITree>({
                collectionName: TREES_COLLECTION_NAME,
                ctx
            });

            expect(res.list).toBeInstanceOf(Array);
            expect(res.list.length).toBe(0);
        });

        test('Should use custom filter condtion supplied', async () => {
            const mockDbServCustom = {
                db: new Database(),
                execute: global.__mockPromise([
                    {
                        _key: 'test_attr',
                        _id: 'core_attributes/test_attr',
                        _rev: '_Wm_Qdtu--_',
                        label: {
                            fr: 'Test'
                        }
                    }
                ])
            };
            testDbUtils = dbUtils({
                'core.infra.db.dbService': mockDbServCustom,
                config: mockConf as IConfig
            });
            testDbUtils.cleanup = jest.fn();
            testDbUtils.convertToDoc = jest.fn().mockReturnValue({
                libraries: ['test']
            });

            const customFilter = jest.fn(() => aql`CUSTOM FILTER`);
            const filters: IAttributeFilterOptions = {libraries: ['test']};
            await testDbUtils.findCoreEntity({
                collectionName: ATTRIB_COLLECTION_NAME,
                filters,
                customFilterConditions: {libraries: customFilter},
                ctx
            });

            expect(customFilter).toBeCalled();
            expect(mockDbServCustom.execute.mock.calls[0][0].query.query).toMatch(/(CUSTOM FILTER){1}/);
        });
    });
    describe('migrate', () => {
        test('Run core migrations', async () => {
            // Mock migration files
            const mockRun1 = jest.fn();
            const file1 = {
                default: () => ({
                    run: mockRun1
                })
            };
            const mockRun2 = jest.fn();
            const file2 = {
                default: () => ({
                    run: mockRun2
                })
            };
            (loadMigrationFile as jest.FunctionLike) = global.__mockPromiseMultiple([file1, file2]);

            // Mock migration files reading
            (readdirSync as jest.FunctionLike) = jest.fn().mockReturnValue(['000.ts', '001.ts']);
            (resolve as jest.FunctionLike) = jest
                .fn()
                .mockReturnValueOnce('/fakeDir/migrations')
                .mockReturnValueOnce('/fakeDir/migrations/000.ts')
                .mockReturnValueOnce('/fakeDir/migrations/001.ts');

            // Mock DB functions
            const mockCollecSave = jest.fn();
            const mockDb = {
                Database: jest.fn(),
                listCollections: global.__mockPromise([]),
                collection: () =>
                    (({
                        create: jest.fn(),
                        save: mockCollecSave
                    } as unknown) as DocumentCollection)
            };
            const mockDbServ: Mockify<IDbService> = {
                db: (mockDb as unknown) as Database,
                execute: global.__mockPromise([])
            };

            const mockDepsManager = {
                build: depDefault => {
                    return depDefault();
                }
            };
            (asFunction as jest.FunctionLike) = m => m;

            const mockLogger: Mockify<Winston> = {
                info: jest.fn()
            };

            const mockPluginsRepo: Mockify<IPluginsRepo> = {
                getRegisteredPlugins: jest.fn().mockReturnValue([])
            };

            const testDbUtils = dbUtils({
                'core.infra.db.dbService': mockDbServ,
                'core.utils.logger': mockLogger as Winston,
                'core.infra.plugins': mockPluginsRepo as IPluginsRepo,
                config: mockConf as IConfig
            });

            await testDbUtils.migrate(mockDepsManager as AwilixContainer);

            expect(mockRun1).toBeCalled();
            expect(mockRun2).toBeCalled();
            expect(mockCollecSave).toBeCalledTimes(2);
        });
    });
});
