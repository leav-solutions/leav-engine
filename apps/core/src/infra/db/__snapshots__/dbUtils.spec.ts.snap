// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dbUtils findCoreEntity Filter with a LIKE on ID 1`] = `"FOR el IN @@value0 FILTER LIKE(el.@value1, @value2, true) FILTER el.@value3 == @value4 FILTER LIKE(el.label.@value5, @value6, true) OR LIKE(el.label.@value7, @value6, true) RETURN el"`;

exports[`dbUtils findCoreEntity Filter with a LIKE on ID 2`] = `
{
  "@value0": "core_trees",
  "value1": "_key",
  "value2": "categories",
  "value3": "system",
  "value4": false,
  "value5": "fr",
  "value6": "Arbre des catégories",
  "value7": "en",
}
`;

exports[`dbUtils findCoreEntity Find core entity without filters 1`] = `"FOR el IN @@value0 RETURN el"`;

exports[`dbUtils findCoreEntity Find core entity without filters 2`] = `
{
  "@value0": "core_trees",
}
`;

exports[`dbUtils findCoreEntity Should filter label on any language 1`] = `"FOR el IN @@value0 FILTER LIKE(el.@value1, @value2, true) FILTER el.@value3 == @value4 FILTER LIKE(el.label.@value5, @value6, true) OR LIKE(el.label.@value7, @value6, true) RETURN el"`;

exports[`dbUtils findCoreEntity Should filter label on any language 2`] = `
{
  "@value0": "core_trees",
  "value1": "_key",
  "value2": "categories",
  "value3": "system",
  "value4": false,
  "value5": "fr",
  "value6": "Arbre des catégories",
  "value7": "en",
}
`;

exports[`dbUtils findCoreEntity Should limit results 1`] = `"FOR el IN @@value0 LIMIT @value1, @value2 RETURN el"`;

exports[`dbUtils findCoreEntity Should limit results 2`] = `
{
  "@value0": "core_trees",
  "value1": 0,
  "value2": 5,
}
`;

exports[`dbUtils findCoreEntity Should sort results 1`] = `"FOR el IN @@value0 SORT el.@value1 @value2 RETURN el"`;

exports[`dbUtils findCoreEntity Should sort results 2`] = `
{
  "@value0": "core_trees",
  "value1": "system",
  "value2": "asc",
}
`;
